File: java/src/java/com/d_project/qrcode/QRNumber.java
package com.d_project.qrcode;

/**
 * QRNumber
 * @author Kazuhiko Arase 
 */
class QRNumber extends QRData {

	public QRNumber(String data) {
		super(Mode.MODE_NUMBER, data);
	}
	
	public void write(BitBuffer buffer) {

		String data = getData();
		
		int i = 0;

		while (i + 2 < data.length() ) {
			int num = parseInt(data.substring(i, i + 3) );
			buffer.put(num, 10);
			i += 3;
		}
		
		if (i < data.length() ) {
			
			if (data.length() - i == 1) {
				int num = parseInt(data.substring(i, i + 1) );
				buffer.put(num, 4);
			} else if (data.length() - i == 2) {
				int num = parseInt(data.substring(i, i + 2) );
				buffer.put(num, 7);
			}

		}
	}
	
	public int getLength() {
		return getData().length();
	}

	private static int parseInt(String s) {
		int num = 0;
		for (int i = 0; i < s.length(); i++) {
			num = num * 10 + parseInt(s.charAt(i) );
		}
		return num;
	}

	private static int parseInt(char c) {

		if ('0' <= c && c <= '9') {
			return c - '0';
		}
		
		throw new IllegalArgumentException("illegal char :" + c);
	}	

	
}

File: ts/src/ts/com/d_project/qrcode/QRNumber.ts
/// <reference path="QRData.ts" />
'use strict';
namespace com.d_project.qrcode {

  /**
   * QRNumber
   * @author Kazuhiko Arase
   */
  export class QRNumber extends QRData {

    constructor(data : string) {
      super(Mode.MODE_NUMBER, data);
    }

    public write(buffer : BitBuffer) : void {

      var data = this.getData();

      var i = 0;

      while (i + 2 < data.length) {
        buffer.put(QRNumber.strToNum(data.substring(i, i + 3) ), 10);
        i += 3;
      }

      if (i < data.length) {
        if (data.length - i == 1) {
          buffer.put(QRNumber.strToNum(data.substring(i, i + 1) ), 4);
        } else if (data.length - i == 2) {
          buffer.put(QRNumber.strToNum(data.substring(i, i + 2) ), 7);
        }
      }
    }

    public getLength() : number {
      return this.getData().length;
    }

    private static strToNum(s : string) : number {
      var num = 0;
      for (var i = 0; i < s.length; i += 1) {
        num = num * 10 + QRNumber.chatToNum(s.charAt(i) );
      }
      return num;
    }

    private static chatToNum(c : string) : number {
      if ('0' <= c && c <= '9') {
        return c.charCodeAt(0) - '0'.charCodeAt(0);
      }
      throw 'illegal char :' + c;
    }
  }
}
